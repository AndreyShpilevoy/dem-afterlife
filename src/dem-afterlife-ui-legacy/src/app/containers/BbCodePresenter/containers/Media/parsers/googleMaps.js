const calculateRoadMapZoomLevel = (value, result = 2) => value >= 2 ? calculateRoadMapZoomLevel(value / 2, result + 1) : result;
const getSatelliteZoomLevel = satelliteZoom => calculateRoadMapZoomLevel(45875200 / satelliteZoom);

const googleMaps = sourceLink => {
    const parsedLink = sourceLink.match(/google(?:\.com)?\.\w+\/maps\/(?:@(\d+\.\d+),(\d+\.\d+),(\d+|\d+.\d+)([zm]))|(embed\?[!=.-_\w\d]+)/i);
    if (parsedLink) {
        if (parsedLink[4] === 'z') {
            return {
                type: 'iframe',
                success: true,
                url: `https://maps.google.com/maps?ll=${parsedLink[1]},${parsedLink[2]}&t=m&z=${Number.parseInt(parsedLink[3], 10)}&output=embed`,
                isShortHeight: false
            };
        } else if (parsedLink[4] === 'm') {
            return {
                type: 'iframe',
                success: true,
                url: `https://maps.google.com/maps?ll=${parsedLink[1]},${parsedLink[2]}&t=h&z=${getSatelliteZoomLevel(parsedLink[3])}&output=embed`,
                isShortHeight: false
            };
        } else if (parsedLink[5]) {
            return {
                type: 'iframe',
                success: true,
                url: `https://www.google.com/maps/${parsedLink[5]}`,
                isShortHeight: false
            };
        }
    }
    return {success: false};
};

export default googleMaps;