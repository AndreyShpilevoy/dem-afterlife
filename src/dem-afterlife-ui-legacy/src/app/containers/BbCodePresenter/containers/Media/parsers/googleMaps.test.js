/* eslint-disable react/jsx-filename-extension, no-undef, fp/no-unused-expression, fp/no-nil, max-statements */

import googleMaps from './googleMaps';

describe('googleMaps', () => {
    it('should return valid result if pass link (map view)', () => {
        expect(googleMaps('https://www.google.com.ua/maps/@50.4810776,30.6050358,15z?hl=ru')).toMatchSnapshot();
    });

    it('should return valid result if pass link (photo view)', () => {
        expect(googleMaps('https://www.google.com.ua/maps/@50.4810776,30.6050358,2885m/data=!3m1!1e3?hl=ru')).toMatchSnapshot();
    });

    it('should return valid result if pass iframe source', () => {
        expect(googleMaps('<iframe src="https://www.google.com/maps/embed?pb=!1m10!1m8!1m3!1d10155.412578773336!2d30.6050358!3d50.4810776!3m2!1i1024!2i768!4f13.1!5e0!3m2!1sru!2sua!4v1506026613652" width="600" height="450" frameborder="0" style="border:0" allowfullscreen></iframe>')).toMatchSnapshot();
    });

    it('should return unsuccessful result if pass wrong string', () => {
        expect(googleMaps('fake')).toMatchSnapshot();
    });
});
