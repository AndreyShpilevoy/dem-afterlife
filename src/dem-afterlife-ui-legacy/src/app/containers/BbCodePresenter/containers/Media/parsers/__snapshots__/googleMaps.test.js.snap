// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`googleMaps should return unsuccessful result if pass wrong string 1`] = `
Object {
  "success": false,
}
`;

exports[`googleMaps should return valid result if pass iframe source 1`] = `
Object {
  "isShortHeight": false,
  "success": true,
  "type": "iframe",
  "url": "https://www.google.com/maps/embed?pb=!1m10!1m8!1m3!1d10155.412578773336!2d30.6050358!3d50.4810776!3m2!1i1024!2i768!4f13.1!5e0!3m2!1sru!2sua!4v1506026613652",
}
`;

exports[`googleMaps should return valid result if pass link (map view) 1`] = `
Object {
  "isShortHeight": false,
  "success": true,
  "type": "iframe",
  "url": "https://maps.google.com/maps?ll=50.4810776,30.6050358&t=m&z=15&output=embed",
}
`;

exports[`googleMaps should return valid result if pass link (photo view) 1`] = `
Object {
  "isShortHeight": false,
  "success": true,
  "type": "iframe",
  "url": "https://maps.google.com/maps?ll=50.4810776,30.6050358&t=h&z=15&output=embed",
}
`;
